File: 백준/Silver/9012. 괄호/괄호.py
for _ in range(int(input())):
    stack = []
    VPS = True
    line = str(input())
    
    for i in line:
        if i == '(':
            stack.append('(')
        elif i == ')':
            if stack:
                stack.pop()
            elif not stack:
                VPS = False
                break
    
    if not stack and VPS:
        print('YES') 
    else: print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main(void){
    int tc;
    cin >> tc;
    
    while(tc > 0){
        tc--;
        string str;
        cin >> str;
        
        stack<char> st;
        string res = "YES";
        for(auto &c: str){
            if(c == '(')
                st.push(c);
            else if(!st.empty() && c == ')')
                st.pop();
            else {
                res = "NO";
                break;
            }
        }
        if(!st.empty())
            res = "NO";
        
        cout << res << endl;
    }
    return 0;
}